### PR TITLE
fix: re-read persona system_prompt at spawn time for non-pack agents

### PR DESCRIPTION
**Pull request by tellaho**
## What changed for users

**Before:** Editing a persona's system prompt or model had no effect on existing agents — even after stopping and restarting them. The only workaround was to delete the agent and recreate it from scratch.

**After:** Persona edits take effect on the very next agent spawn. Stop an agent, tweak its persona's instructions, start it again — it picks up the new prompt immediately. This makes agent iteration and improvement loops dramatically faster.

> Pack-based agents already had this behavior. This fix brings non-pack persona-linked agents to parity.

## Technical changes

- **Extracted `resolve_effective_prompt_and_model()`** — a pure function that resolves the effective system prompt and model given a `persona_id`, the persona store, and the record's stale snapshot values. No `AppHandle` dependency, trivially testable.
- **`spawn_agent_child()` now re-reads the persona store at spawn time** via `load_personas()` then delegates to the extracted helper. Falls back to `record.system_prompt` / `record.model` when no persona is linked or the persona can't be found (graceful degradation).
- **Extracted tests to `runtime/tests.rs`** following the `env_vars/tests.rs` pattern, bringing `runtime.rs` under the file-size linter limit.
- **Lowered file-size override** from 1110 → 880 in `check-file-sizes.mjs`.

## Testing

- 5 new unit tests covering every branch of `resolve_effective_prompt_and_model`:
  1. `persona_id` matches → reads fresh prompt + model from persona store
  2. No `persona_id` → falls back to record snapshot
  3. `persona_id` not found (deleted) → graceful fallback to record snapshot
  4. No persona and no record prompt → returns None/None
  5. Persona with no model → returns None (doesn't inherit stale record model)
- `cargo test --manifest-path desktop/src-tauri/Cargo.toml --lib` → **354/354 pass**
- `rustfmt --check` ✅
- No schema migrations, no new deps, easy single-commit rollback

image

👆 _Set the agent's instructions from `only respond with 5 words, starting with the letter P`, sent a message, stopped it, set the instructions to `only respond with 5 words, starting with the letter B`, then respawned it._

## Note for the team

The desktop Tauri crate has 354 unit tests (including 5 new ones from this PR), but `just test-unit` and `just ci` don't run them — only `cargo test --manifest-path desktop/src-tauri/Cargo.toml --lib` exercises them locally. Consider wiring this into the CI gate separately.

### CODE REVIEW

**Review by wesbillman — APPROVED** [review]

**Comment by wesbillman** [comment]
@codex please review

File: desktop/src-tauri/src/managed_agents/runtime.rs [hunk 1]
```diff
@@ -500,6 +500,37 @@ pub(crate) fn build_respond_to_env(
...
+) -> (Option<String>, Option<String>) {
+    let from_persona: Option<(String, Option<String>)> = persona_id.and_then(|pid| {
+        personas
+            .iter()
+            .find(|p| p.id == pid)
+            .map(|p| (p.system_prompt.clone(), p.model.clone()))
+    });
+
+    match from_persona {
+        Some((prompt, model)) => (Some(prompt), model),
```
Comment (wesbillman) [inline desktop/src-tauri/src/managed_agents/runtime.rs]: 🤖 This makes the persona prompt/model always win whenever `persona_id` resolves, but the existing app still treats `ManagedAgentRecord.system_prompt` and `.model` as mutable per-agent settings. `EditAgentDialog` persists `systemPrompt`, and `ModelPicker` persists `model`; after this change those saved values become no-ops for persona-linked agents. In particular, if a persona has `model: None`, selecting a model on that agent will still save/display but spawn will remove `SPROUT_ACP_MODEL`. Can we either preserve explicit agent override precedence, or update the UI/data model so persona-backed agents cannot appear to have per-agent prompt/model overrides?

File: desktop/src-tauri/src/managed_agents/runtime.rs [hunk 2]
```diff
@@ -623,30 +654,16 @@ pub fn spawn_agent_child(
...
-                })
-                .map(|p| (p.system_prompt, p.model))
-        })
-        .flatten();
-
-    let (effective_prompt, effective_model) = match persona_prompt_and_model {
-        Some((prompt, model)) => (Some(prompt), model),
-        None => (record.system_prompt.clone(), record.model.clone()),
-    };
+    // Resolve system prompt and model via the extracted helper. Re-reads
```
Comment (wesbillman) [inline desktop/src-tauri/src/managed_agents/runtime.rs:672]: 🤖 This re-read is only wired into local `spawn_agent_child()`. Provider-backed starts/redeploys still build their payload from `record.system_prompt` and `record.model` in `build_deploy_payload()`, so persona edits will not propagate there. If provider agents are in scope, this effective prompt/model resolver should be shared with the deploy payload path too; if not, the PR description should call out that the fix is local-only.

File: desktop/src-tauri/src/managed_agents/runtime.rs [hunk 3]
```diff
@@ -500,6 +500,37 @@ pub(crate) fn build_respond_to_env(
     Ok((set, remove))
 }
 
+/// Resolve the effective system prompt and model for a managed agent.
+///
+/// When `persona_id` is set, looks up the persona in the provided list and
+/// returns its current `system_prompt` and `model`. Falls back to the
+/// agent record's stale snapshot (`record_prompt`, `record_model`) when:
+/// - `persona_id` is `None`
+/// - the persona can't be found in the store (e.g. deleted)
```
Comment (wesbillman) [inline desktop/src-tauri/src/managed_agents/runtime.rs]: 🤖 This fallback is true for the pure helper, but full spawn still calls `resolve_persona_env(app, record.persona_id.as_deref())?` later, which fails closed when `persona_id` is set and missing. The test/description around the missing-persona case may overstate runtime behavior unless the env resolution path is also adjusted.

**Review by wesbillman — COMMENTED** [review]
🤖 Review notes from my pass:

I think the user-facing goal is worth doing: persona edits should take effect after an agent restart. The current implementation looks too broad, though, because it changes the precedence semantics for all persona-linked local agents and leaves provider deploy on the old behavior.

Main things to address:
- Preserve or intentionally remove per-agent prompt/model overrides for persona-linked agents. Right now the UI still saves them, but local spawn ignores them when `persona_id` resolves.
- Apply the same effective prompt/model resolution to provider deploys, or document that this fix is local-only.
- Tighten the missing-persona test/description so it does not imply full spawn succeeds with a missing persona; `resolve_persona_env()` still fails closed later in `spawn_agent_child()`.

I could not complete the local targeted Tauri unit run because the build failed before tests on a missing sidecar resource: `desktop/src-tauri/binaries/sprout-aarch64-apple-darwin`.

File: desktop/src-tauri/src/managed_agents/runtime.rs (hunk 1 above)
Comment (tellaho) [reply desktop/src-tauri/src/managed_agents/runtime.rs]: Resolved in `2c5d721c` — implemented Option A precedence: agent-record values now win when explicitly set (via Edit Agent dialog). Persona values are only used as fallback when the agent record fields are `None`. The migration in `7f7fa99f` clears stale snapshots that match the persona's current defaults, so only genuine user overrides survive.

File: desktop/src-tauri/src/managed_agents/runtime.rs (hunk 2 above)
Comment (tellaho) [reply desktop/src-tauri/src/managed_agents/runtime.rs:672]: Resolved in `e29c895c` — wired `resolve_effective_prompt_and_model` into `build_deploy_payload()` so provider-backed agents also pick up live persona values (with the same Option A precedence: agent override wins when set).

File: desktop/src-tauri/src/managed_agents/runtime.rs (hunk 3 above)
Comment (tellaho) [reply desktop/src-tauri/src/managed_agents/runtime.rs]: Resolved in `e29c895c` — tightened the missing-persona test. The helper now returns `(None, None)` when the persona is missing (no stale fallback), and the test documents that `resolve_persona_env` in the full spawn path fails closed independently. Both paths are consistent: missing persona = no prompt/model injected.

File: desktop/src-tauri/src/managed_agents/runtime.rs [hunk 4]
```diff
@@ -500,6 +500,37 @@ pub(crate) fn build_respond_to_env(
...
+    let from_persona: Option<(String, Option<String>)> = persona_id.and_then(|pid| {
+        personas
+            .iter()
+            .find(|p| p.id == pid)
+            .map(|p| (p.system_prompt.clone(), p.model.clone()))
+    });
+
+    match from_persona {
+        Some((prompt, model)) => (Some(prompt), model),
+        None => (
```
[the remark this answers is not in the document]
Comment (tellaho) [reply desktop/src-tauri/src/managed_agents/runtime.rs]: Resolved in `2c5d721c` — Option A precedence implemented. Agent-record `system_prompt`/`model` now take priority over persona values when explicitly set. The Codex bot's concern (saved per-agent model being ignored) is fully addressed: user overrides via Edit Agent / ModelPicker are respected at spawn time.

**Comment by tellaho** [comment]
## Changes after review feedback

All review comments have been addressed. Here's the full summary:

---

### What was requested (from review comments):

1. **Preserve per-agent model/prompt overrides** (codex-bot P1 + @wesbillman) — persona values were unconditionally winning, making Edit Agent / ModelPicker saves no-ops for persona-linked agents.
2. **Wire persona resolution into provider deploys** (@wesbillman line 673) — the re-read was local-only; provider-backed agents still used stale `record.system_prompt`/`record.model`.
3. **Tighten missing-persona fallback behavior** (@wesbillman) — the helper's fallback description overstated what actually happens at runtime when `resolve_persona_env` fails closed.

---

### What we shipped (4 commits on top of the original fix):

| Commit | What it does |
|--------|-------------|
| `e29c895c` | **fix: apply persona resolution to provider deploys + tighten missing-persona test** — shares `resolve_effective_prompt_and_model` with `build_deploy_payload()` so provider agents get live persona values too. Tightens the missing-persona test to return `(None, None)` instead of stale fallback. |
| `2c5d721c` | **feat: implement Option A precedence — agent-level overrides win over persona** — when agent record has explicit `system_prompt`/`model` (set via Edit Agent), those values take priority. Persona values are fallback only when fields are `None`. |
| `7f7fa99f` | **fix: migrate existing persona-backed agents to clear stale prompt/model snapshots** — one-shot migration clears `system_prompt`/`model` only when the stored value matches the persona's current default. Genuine user overrides survive. 5 unit tests. |
| `a48086b2` | **perf: move persona-defaults migration to one-shot startup path** — moves migration out of `load_managed_agents` (~20+ call sites) into `restore_managed_agents_on_launch` Phase A (once per app launch). `load_managed_agents` is now pure load-and-parse with zero side effects. |

---

### Result:
- All review comments resolved (replied inline with commit SHAs)
- 361+ tests pass, fmt + clippy clean
- No format changes, no new files, idempotent on crash
- `runtime.rs` file-size override bumped 880→915 to accommodate expanded Option A doc comments

**Comment by tellaho** [comment]
deleting in favor of some upcoming iterations between @wpfleger96 and me
